### PR TITLE
Allow knocking rooms

### PR DESCRIPTION
**Pull request by charlynguyen**
We @nordeck are currently implementing the knock rooms behind the feature flag `feature_ask_to_join` (introduced in https://github.com/matrix-org/matrix-react-sdk/pull/11182).

This pull request allows knocking rooms.
- Request to join (a room)
- Cancel request (to join a room)
- Center aligns the spinner

#### ~Is blocked by~

- ~https://github.com/matrix-org/matrix-js-sdk/pull/3647~

#### Relates to

- https://github.com/matrix-org/matrix-react-sdk/pull/11248
- https://github.com/matrix-org/matrix-react-sdk/pull/11182
- https://github.com/vector-im/element-web/pull/25923

### Screens / Videos

| Request to join | Cancel request | Spinner |
|-|-|-|
|request-to-join|cancel-request|spinner|

#### Guess knock

The video shows the user flow for rooms where we need to guess if we can knock. It is agreed on with @daniellekirkwood and @callumu.

https://github.com/matrix-org/matrix-react-sdk/assets/1422657/e819acd6-9100-4460-8ce7-8420cf84067e

#### Happy path

https://github.com/matrix-org/matrix-react-sdk/assets/1422657/5d5b9a78-6199-4e36-84d0-fe4fc4f09cca

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Allow knocking rooms ([\#11353](https://github.com/matrix-org/matrix-react-sdk/pull/11353)). Contributed by @charlynguyen.

### CODE REVIEW

File: src/dispatcher/actions.ts [hunk 1]
```diff
@@ -351,4 +351,19 @@ export enum Action {
...
+
+    /**
+     * Fired when requesting to ask to join a room.
+     */
+    AskToJoin = "ask_to_join",
+
+    /**
+     * Fired when requesting to submit an ask to join a room. Use with a SubmitAskToJoinPayload.
+     */
+    SubmitAskToJoin = "submit_ask_to_join",
```
Comment (germain-gg) [inline src/dispatcher/actions.ts:363]: What difference is there between the two? 

`AskToJoin` is ambiguous, "fired when request to ask to join a room" implies that a request has been submitted. But the second action below contradicts that.

Could we clear that up?

File: src/dispatcher/payloads/CancelAskToJoinPayload.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
```
Comment (germain-gg) [inline src/dispatcher/payloads/CancelAskToJoinPayload.ts:2]: Incorrect copyright header

File: src/dispatcher/payloads/SubmitAskToJoinPayload.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
```
Comment (germain-gg) [inline src/dispatcher/payloads/SubmitAskToJoinPayload.ts:2]: Incorrect copyright header

File: src/components/structures/RoomView.tsx [hunk 4]
```diff
@@ -1991,6 +2010,29 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
         );
     }
 
+    private onSubmitAskToJoin = (reason?: string): void => {
```
Comment (germain-gg) [inline src/components/structures/RoomView.tsx:2013]: Can we add a JSDoc comment, same comment for the following functions in this file

File: src/stores/RoomViewStore.tsx [hunk 5]
```diff
@@ -746,4 +770,35 @@ export class RoomViewStore extends EventEmitter {
     public isViewingCall(): boolean {
         return this.state.viewingCall;
     }
+
+    public askToJoin(): boolean {
```
Comment (germain-gg) [inline src/stores/RoomViewStore.tsx:774]: Can we add a JSDoc comment? Same comment for the function after in this file.

**Review by germain-gg — CHANGES_REQUESTED** [review]
Thank you for your contribution!

Overall looks very good to me, with some minor changes requested

File: src/components/structures/RoomView.tsx [hunk 6]
```diff
@@ -395,6 +402,8 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
     public constructor(props: IRoomProps, context: React.ContextType<typeof SDKContext>) {
         super(props, context);
 
+        this.askToJoinEnabled = SettingsStore.getValue("feature_ask_to_join");
```
[the remark this answers is not in the document]
Comment (charlynguyen) [reply src/components/structures/RoomView.tsx:405]: https://github.com/vector-im/element-web/pull/25923

File: src/dispatcher/payloads/CancelAskToJoinPayload.ts (hunk 2 above)
Comment (charlynguyen) [reply src/dispatcher/payloads/CancelAskToJoinPayload.ts:2]: Sorry 😅

File: src/dispatcher/payloads/SubmitAskToJoinPayload.ts (hunk 3 above)
Comment (charlynguyen) [reply src/dispatcher/payloads/SubmitAskToJoinPayload.ts:2]: Thanks for the catch!

File: res/css/views/rooms/_RoomPreviewBar.pcss [hunk 7]
```diff
@@ -148,3 +149,12 @@ a.mx_RoomPreviewBar_inviter {
     text-decoration: underline;
     cursor: pointer;
 }
+
+.mx_RoomPreviewBar_icon {
+    margin-right: 8px;
+    vertical-align: text-top;
+}
+
+.mx_RoomPreviewBar_full {
```
[the remark this answers is not in the document]
Comment (charlynguyen) [reply res/css/views/rooms/_RoomPreviewBar.pcss]: I changed it to `.mx_RoomPreviewBar_fullWidth`. I have seen you already used `…fullWidth` in your codebase and it is given more context 🙂

File: src/dispatcher/actions.ts (hunk 1 above)
Comment (charlynguyen) [reply src/dispatcher/actions.ts]: For the "guess knock" (see description) I need away to bring the form up for the user. Maybe the naming is not the best. I changed it to `PromptAskToJoin`. I hope it make the difference clearer.

**Review by germain-gg — APPROVED** [review]
Looks good to me!

Minor thing regarding the review, but could you avoid force pushing? The commits ends up being squashed, and by force pushing you erase the review history "changed since last viewed", thank you 🙏

**Review by daniellekirkwood — COMMENTED** [review]
It's already merged but FWIW I approve #product :) 

Thank you!

**Pull request merged**